### PR TITLE
Add test result for identifier page user creation part

### PR DESCRIPTION
**Pull request by oliverviljamaa**
## Background

We'll be adding an _add test result for identifier page_ that allows adding a test result for an identifier and creates a user for it in the background. 

## Changes

The **user creation part of this page** has been added.

The PR can be merged as is, as it's not breaking anything, or when the test part is implemented, to be feature-complete. Opened the PR now to get feedback sooner.

Some client-side validation and different copy based on the authentication method of the deployment has been added.

### CODE REVIEW

File: src/admin/BulkUserCreationPage.tsx [hunk 1]
```diff
@@ -33,7 +33,7 @@ const BulkUserCreationPage: FC = () => {
     onSubmit: ({ role, emailsString }) => {
       const emails = getEmails(emailsString);
       // TODO: support ESTONIAN_ID authenticaiton method
-      const command: CreateUserCommand[] = emails.map((email) => ({
+      const command: CreateUserWithRolesCommand[] = emails.map((email) => ({
```
Comment (oliverviljamaa) [inline src/admin/BulkUserCreationPage.tsx:36]: Renaming to not clash with the new `CreateUserCommand` and to be more explicit.

File: src/App.tsx [hunk 2]
```diff
@@ -62,6 +62,13 @@ const App = () => {
       <AuthenticatedRoute path="/scan" exact>
         <ScanPage />
       </AuthenticatedRoute>
+      <AuthenticatedRoute
+        path="/add-test"
+        exact
+        requiredPermissions={['CREATE_USERS']} // TODO: Add 'CREATE_TESTS_WITHOUT_ACCESS_PASS' when test adding is there
```
Comment (StevePole) [inline src/App.tsx:68]: It's merged now

File: src/testing/AddTestToIdentifierPage.tsx [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+import * as yup from 'yup';
+
+import { CreateUserCommand, AuthenticationMethod } from '../api';
+import { useConfig } from '../common';
+import { useUserCreation } from './useUserCreation';
+
+const AnyBox = Box as any;
+
+// from https://ipsec.pl/data-protection/2012/european-personal-data-regexp-patterns.html
+const ESTONIAN_ID_CODE_REGEX = /^[1-6][0-9]{2}[1,2][0-9][0-9]{2}[0-9]{4}$/;
```
Comment (oliverviljamaa) [inline src/testing/AddTestToIdentifierPage.tsx:14]: I double-checked their regex and I think it's wrong — `[1,2]` is the first digit of `MM`, meaning it wouldn't allow `01`, for example 🤔

I'll replace with the `isikukood` package to have the same validation as the server.

File: src/App.tsx (hunk 2 above)
Comment (oliverviljamaa) [reply src/App.tsx:68]: Yup, at the moment this PR is just not adding the tests yet, so not really using this permission 👍

File: src/admin/BulkUserCreationPage.tsx [hunk 4]
```diff
@@ -32,8 +32,8 @@ const BulkUserCreationPage: FC = () => {
     }),
     onSubmit: ({ role, emailsString }) => {
       const emails = getEmails(emailsString);
-      // TODO: support ESTONIAN_ID authenticaiton method
-      const command: CreateUserCommand[] = emails.map((email) => ({
+      // TODO: support ESTONIAN_ID authentication method
```
Comment (StevePole) [inline src/admin/BulkUserCreationPage.tsx:35]: Is this still relevant?  What needs to be added?

File: src/common/useConfig.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import { Config, AuthenticationMethod, Language } from '../api';
+
+export function useConfig(): Config {
```
Comment (StevePole) [inline src/common/useConfig.ts:3]: Can you make this an async function so that you won't need to change the consumers later?

File: src/testing/AddTestToIdentifierPage.test.tsx [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+import nock from 'nock';
+
+import { useAuthentication } from '../authentication/context';
+import { renderWrapped } from '../testHelpers';
+import { AddTestToIdentifierPage } from '.';
+
+jest.mock('../authentication/context');
+
+describe(AddTestToIdentifierPage, () => {
+  it('creates a user for the authentication method from config and the filled identifier', async () => {
```
Comment (StevePole) [inline src/testing/AddTestToIdentifierPage.test.tsx:12]: Any way to break this test up into smaller, more documented pieces?  It's quite hard to follow at the moment.

File: src/admin/BulkUserCreationPage.tsx (hunk 4 above)
Comment (oliverviljamaa) [reply src/admin/BulkUserCreationPage.tsx:35]: The bulk user creation page currently only works for the `MAGIC_LINK` method indeed, as it hardcodes the method and has email-specific copy.

When we see a use case for the page with other methods, we can mimic what's being done in this PR for the add test for identifier page and change copy and passed authentication method depending on the config.

File: src/common/useConfig.ts (hunk 5 above)
Comment (oliverviljamaa) [reply src/common/useConfig.ts:3]: It's a React hook for usage in components, so for the consumers, it'll be a sync function 👍

File: src/testing/AddTestToIdentifierPage.test.tsx (hunk 6 above)
Comment (oliverviljamaa) [reply src/testing/AddTestToIdentifierPage.test.tsx:12]: Yup, will try to improve without creating too much repetition.

Do you feel comments and/or well named helper functions would help?

Comment (StevePole) [reply src/testing/AddTestToIdentifierPage.test.tsx:12]: I like to use BDD, pseudocode:
```
describe('when initialised', () => {
  beforeEach(() => {
    mount(...);
  });
  it('should show stuff ', () => { ... }

  describe('when text is entered the user clicks submit', () => {
    let promise;
    beforeEach(() => {
      button.click();
      promise = new Promise()
      mock(api.method).returnValue(promise)
    });
    it('should make a request to the xxx API', () => { ... });

    describe('when the request fails', () => {
      beforeEach(() => {
        promise.reject(...);
      });
      it('should show an error', () => { ...});
    });

    describe('when the request succeed', () => {
      beforeEach(() => {
        promise.resolve(...);
      });
      it('should ...', () => { ...});
    });
  });
});
```
A bit more verbose, but helps me lay stuff out in a readable way.  Don't feel obliged to follow that.

Comment (oliverviljamaa) [reply src/testing/AddTestToIdentifierPage.test.tsx:12]: I've followed BDD as well, but for React Testing Library's integration tests, I've started writing longer isolated tests and have found them to be more maintainable thanks to the immutability and isolation, especially with hooks and async operations.

Although slightly exaggerated, I agree with the main points from Kent C. Dodds:
https://kentcdodds.com/blog/write-fewer-longer-tests
https://kentcdodds.com/blog/avoid-nesting-when-youre-testing

That being said, a test definitely shouldn't lose its readability, so I'll figure out a way to improve 👍

Comment (oliverviljamaa) [reply src/testing/AddTestToIdentifierPage.test.tsx]: ✅ Used `beforeEach` for the setup (mocking the token), added a `fillInput` helper function, and split apart testing the autofocus. When I went further, I started sacrificing readability, flow integration confidence, and maintainability.

File: src/common/useConfig.ts (hunk 5 above)
Comment (Tankenstein) [reply src/common/useConfig.ts:3]: Still we need to pass some concept of loading that the components can handle imo, or just return null

Comment (oliverviljamaa) [reply src/common/useConfig.ts:3]: Yup, I'd expect it to return `null` when not initialised.
